### DISCUSSION TITLE
why vscode is a better solution for remix desktop

### DISCUSSION THREAD

**Discussion by drafish** [PlanckerLabs/remix-cc#96] [Ideas]
## Why vscode is a better solution
I think we both agree that the current remix desktop is not so desktop. I read the code of remix desktop, it's simply just load the web version of remix. The only difference between desktop and web is that desktop contains remixd and cannot use metamask. In this case, I prefer web rather than desktop.

It's good to see that remix team is working on new desktop which will contains real terminal and native git. But I wonder how many resources you're gonna put on the new desktop. 

As far as I know, when it comes to desktop, vscode is the first choice. Especially for the senior solidity developers, they don't like to use remix on browser, because it's limited. People will expect remix new desktop to be as good as vscode. Otherwise they wouldn't switch their choice. It will cost massive work for remix new desktop to achieve vscode level.

Besides, remix new desktop may cause another problem. I notice that there is a PR called [Remix Desktop Alpha](https://github.com/ethereum/remix-project/pull/3885) in remix-project codebase, and there is a lot `isElectron` in it. I assume that the new desktop will be in the same codebase with web. The codebase is already bloated, it will be a disaster if it need to be compatible with both web and desktop.

In my point of view, it will be better if we keep the codebase clean for web and find another solution for desktop. By another solution, I mean vscode. Because vscode is popular and opensource. We don't need to develop file explorer and terminal and git etc, because vscode got them all covered. All we need to do is making it adapted to solidity and other smart contract languages. Besides, there are other advantages of vscode that we can take, such as github codesapces —— cloud-based development environment. How you gonne manage that in remix if you do it on your own?

## How do we start 
We can start from a new remix vscode extension. I'm thinking about the way that `Extension Pack for Java` does —— a big extension which contains several sub extensions. We can call it `Extension Pack for Remix`. 

For the first version, we can just implement three function modules.
1. Solidity Compiler
2. Deploy & Run Transactions
3. Debugger

We don't need to do it all by ourselves. Because there are two extensions that sort of implement these features.
1. remix-light is a light version of `Solidity Compiler` and `Deploy & Run Transactions`. —— https://github.com/arthurshoppy/remix-light
2. solidity-debugger brings native debugging support for the popular Hardhat framework. —— https://github.com/robertaachenw/solidity-debugger

All we need to do is put them together and enhance them, and make it easier for users to use.

We can contact @arthurshoppy and @robertaachenw to see if they are interested to develop the extension together.

For further version, we can add more sub extensions, such as `solidity unit testing` and `solidity analyzers`, and add support for other smart-contract languages.

## Maybe even replace remix web version
VSCode got web version. Although there are limitations and extension may not work, we can do some deep customization to make it work. Besides, it works fully on remote and cloud-based development environment.

It's just some immature thoughts. I'm not sure how remix team feels about that. To me, it's like edge. Microsoft gave up the old edge, and used chromium to develop the new edge. It's exactly the same situation. @bunsenstraat

**Comment by bunsenstraat** [comment]
Thanks! This opinion here is just mine, the other members of the Remix team may think otherwise.

To start, we already spend considerable time and resources on developing and releasing a VSCode extension which actually featured a debugger, a run and deploy app, a solidity compiler, unit testing, support for plugins, a bridge to the web IDE to use MetaMask and remixd. 
You could summarise this effort as: workarounds, compromise, watering down, not like Remix at all and hard to maintain. 
You say: 'It will cost massive work for remix new desktop to achieve vscode level.' 
I say: it will cost massive work for a VScode extension to achieve Remix level.

VScode extensions are super great if you build something in there from the ground up, designing the app with the restrictions in mind that comes with vscode extensions. Most of this comes down to very limited UI & internal API, security, communication between components. You have to deal with what you get, not with what you design. And what you get changes without having any control over it.

This is all fine in itself and very useful, but to adapt the current Remix UI and the interaction between all the components that Remix has is a huge pain. This integration between tools is basically what Remix is and makes it work. It's not just an editor and filesystem with some tools added to it. 
We tried work around that by loading separate web apps that were limited versions of our live modules, ie the compiler, run & deploy, debugger etc into the web views of VSCode. They were Remix plugins and could talk to each other and the base extension. But because you don't have any control over the real estate in VSCode it was very hard to fit all of this in. It worked but performance was bad, UX was limited and the app crowded, and most of all the functionality was very very restricted. Not nice at all. It was not like Remix.

Rebuilding all of those tools the way Remix wants them to be into the limitations of a VSCode extension was not a viable solution at the time. You basically need to build everything again. You can do that but you need to remember how much detail is built into Remix on how everything works together to create the workflow it has. This a huge, insane amount of work and it would not result in coming close to what Remix now offers.

CI & E2E. Maybe this has changed but at the time there seemed to be no way to have CI or E2E for your extension. You basically submit a build once in while and they run tests on it. Remix, due to its complexity and huge amount of little tiny details relies very heavily on e2e testing. It's impossible to manually test Remix. The e2e is an app in itself. Without that and proper CI, it is unmaintainable. 

Dependancy. IMO building a product entirely on a big platform like VSCode you are setting yourself up to always catching up with any random changes they may make to say extension security, UI changes, API changes etc etc. And these may change due to VSCode being a massive product that's used for all kinds of stuff, not only to support your code. A feature that you successfully worked around to make something work the way you want at one time might be made completely impossible the next release of VSCode. You have no control whatsoever. This is again not uncommon.

As to Remix Desktop Alpha, sure it's a work in progress and this a POC with fast and sloppy code. Make it work and test it. Performance is big issue, but at least we have complete control of it. VSCode is an electron app too and they developed their own codebase using no dependency on an UI library. They made it fast and efficient on resources. That's great. We use React, we might fall on our face when React is no longer maintained. All true. But we can do whatever we want in our desktop app without compromising anything. If people want to edit in VSCode, that's perfectly possible, the changes will be visible in Remix and you can use the tools we provide there to do your web3 stuff. I'm convinced we can make Remix desktop perform great. We also maintain our codebase and run our own CI and E2E to ensure the experience for the user is consistent. We can integrate advanced features that require a desktop environment. Again with full control.  
Saying a codebase is bloated doesn't mean anything unless it makes our app less efficient, which it doesn't. It works well. It should perform better in low bandwidth situations, that is for sure. And fully support offline use. Code can be optimised and cleaned up always, of course. Many things can be improved always.

All of this said, there is enough space for devs to create tools for VSCode, for dektop or web versions, to do stuff that Remix does. All the better! More tools for all styles. 
Using existing vscode extensions to create a pack of extensions that deal with web3 stuff in vscode is great, and it should be done for sure. 
But it's not Remix. It's what is says, another collection of great extensions. But you also fight for space with other extensions and there's no way at all you can influence that. Technical conflicts with extensions are common, look at Solidity and Hardhat together. In Remix we can make stuff happen all over in whatever way we want. Complete control.
